### ISSUE TITLE
plugin unit tests should use directories beneath _trial_temp, not twisted/plugins/

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10632]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#942|
|Type|defect|
|Created|2005-03-21 21:10:17Z|
|Branch|https://github.com/twisted/twisted/tree/test-plugins-not-shared-942|

<details><summary>Searchable metadata</summary>

```
trac-id__942 942
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_test_plugins_not_shared_942 branches/test-plugins-not-shared-942
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1111439417000000 1111439417000000
changetime__1195587186000000 1195587186000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__spiv cc__itamarst cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
twisted/plugins/ is not always writeable.  skipping the tests is silly.  if we
want to test behavior when write permission is denied, we can use os.chmod().
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [21759]) Branching to 'test-plugins-not-shared-942'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21761]) Fix the behavior, add some docstrings.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

This is ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set owner to @therve|
|-|-|

Hooray for making things more testable!

`test_plugins` in `test_plugin.py` has docstring quotes but doesn't have any actual docstring.

`_testDetectNewFiles` is big, confusing and has obscure intent. Similarly for `_testDetectFilesRemoved`.

The ticket would have been easier to review if you included a comment on the ticket explaining how you solved the problem.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [21775]) Add some docs and comments.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

I guess you understand how I solved the problem, but I'll explain anyway for the record.

Instead of writing files directly in twisted.plugins, I created them in a temporary directory, and then added this directory to ```plugins.__path__```. This way getPlugins is able to find it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

I haven't had a chance to look at the code, but that makes it sound like the problem is only half-solved.  Ideally (and I thought that's what this ticket meant), tests shouldn't be looking at twisted.plugins at all, they should be creating a new package just for testing.  (_Really_ ideally, twisted.plugin would also be refactored such that it could be an object, e.g. `PluginRegistry`, that didn't need to work on the real `sys.path` at all, but maybe that's a bigger and probably separate task.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

Well the version in the branch doesn't modify twisted.plugins anymore, but it does still use it as main entry point, yes. 

To be clear, my first intent was to fix the tests for make it pass under disttrial, which I did. If the expectations are higher, fine, I'll probably be able to (try to) fix it too. If I understand correctly, we should not use twisted.plugins at all, and use another directory instead?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml set owner to @therve|
|-|-|

Out of everything Glyph said, this bit is the most important:
  "Ideally, tests shouldn't be looking at twisted.plugins at all, they should be creating a new package just for testing."

I've just looked at the diff again. The extra comments definitely help (thanks!), but I still don't really understand what's going on.

`test_plugins` calls `_testWithCacheness` (what does that do?) which appears to call `_testPlugins` twice, call `_clearCache` and then call `_testPlugins` twice again. `_testPlugins` gets the plugins for `ITestPlugin2` which doesn't seem to be set up anywhere. 

Oh! It's defined in `twisted/plugin.py`. Yikes. I am very surprised that this patch was allowed in -- I remember it as being quite recent. This code is most definitely not your fault, therve, so please don't take offense.

I'm happy for you to merge your changes in, but it definitely does not fix this ticket. IMO, for this ticket to be fixed, _all_ of the test code must be moved out of `twisted/plugin.py` and `twisted/plugins/`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Mea culpa.  The code in question is definitely my fault: I can remember writing it.  In Mantissa.  The *old* Mantissa.  In other words, it wasn't that recent.

jml, you were probably thinking of r19305, which was only 10 months ago.  The ill-conceived test interfaces, however, have been with twisted.plugin since the beginning - r13168, over 3 years ago.

It is because of the long, grim shadow of those dark interfaces that I was interested in the comments on this ticket.  _They must die._  I will spare you the scary stories of all the horrible stuff that those tests have done in the past.  Suffice it to say that they're every bit as bad as they look.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Oh, and to be clear: therve, thank you for the work you have done so far.  I didn't mean to say that the changes shouldn't be merged either.  Only to highlight that there was more of a problem.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

(In [21794]) It don't work anymore but I need to clean this up

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

(In [21795]) Fix it!

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve removed owner|
|-|-|

I think it deserves a first look.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

  * A more reliable way for ```PluginTestCase``` to manage ```sys.modules``` and ```sys.path``` is shown by ```AdjacentPackageTests```.
  * The whole test module might be a little nicer to look at if it didn't have big strings lying all over it.  The contents could be in files in the test directory and the tests could copy them around.  This probably isn't really a blocking issue for this ticket though, since it's just an aesthetic thing and the branch isn't introducing it, just adding more of it.
  * Some test coverage almost existed for the permission denied cases (ie, the "testing in deployed mode" crap).  This change gets rid of it completely.  We should probably add real tests for those cases at some point.  Perhaps for a separate ticket.
  * ```_unimportPythonModule``` changed to delete from ```sys.modules``` before deleting the package attribute, instead of doing it the other way around.  I can't figure out why, though. :)  Could you explain it, at least here?  A comment about the order in the code might be appropriate too (but maybe it depends what the explanation is).

The tests are still kind of bad overall, I think, but this is a big step in a great direction.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

(In [21810]) Create new files for test plugins, change way of saving and restoring sys.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

(In [21811]) Add a new test for deployed mode.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve commented|
|-|-|

(In [21812]) Screw you windows

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve removed owner|
|-|-|

Ok I think I've addressed all your points. The new test isn't as good as I hoped, but windows is really bad, so that's the best I could get for now. I open for proposals though.

Regarding your last point, it was just a mistake, I reverted that.

Back to review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_27"></a>|@jml set owner to @zooko|
|-|-|

Here's my review as a non-expert in `easy_install` & friends.

`twisted/python/dist.py`
 * Need spaces between operators (see PEP 8) in build_ext_no_fail
 
`setup.py`
 * Needs a module docstring.
 * Needs a docstring for `main()`
 * Duplicates significant amounts of data from `twisted.topfiles.setup`. It would seem more sensible to copy the dict and mutate the values, or at least refer to values in the topfiles `setup_args` dict.

I also can't find out where `setupdist.py` is used. Grep doesn't help me.

I also think this branch needs RM approval before landing.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_28"></a>|@jml removed owner|
|-|-|

OOPS!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_29"></a>|@jml set owner to @therve|
|-|-|

So glad you deleted that crap from twisted.plugins.

`test_plugin.py`
 * Obviously much cleaner. Thanks.
 * `_testWithCacheness` is still confusing and weird and unexplained.
 * The docstring for `test_cache` doesn't add anything to the test. It should explain what is expected and (ideally) why.
 * _testCache should explain what it is doing. I think it would be clearer if it were parametrized and renamed to something like `assertPluginInCache` (assuming that's what it actually does).
 * `test_plugins` should explain why it's important that the instances have `test` methods and where this comes from
 * `test_plugins` seems to be similar to `test_cache` -- it's hard to figure out what it's doing.

It's worth a digression here. In general, unit tests have four parts: setting up the fixtures, exercising the system being tested, making assertions about the state or behaviour of the system,  and cleaning up anything that needs it.

Looking at `test_cache` and `test_plugins`, the set up and clean up are clear, but it is not clear which parts of the system are being exercised, and the assertions appear scattered.

This is partly because of the `testWithCacheness` weirdness which moves the test out of the actual test code. We can probably get rid of this with a decorator.

 * `_testDetectNewFiles` should use `addCleanup` instead of a try/finally. Oh, wait, no you can't, because the test is run four times. Ugh! So weird.


That's it. Sorry to write so much -- I know you aren't responsible for most of these tests. Perhaps much of what I say can be deferred to another ticket.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_30"></a>|@therve commented|
|-|-|

(In [21874]) Add some docs.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve commented|
|-|-|

(In [21875]) Maybe this is simpler.

Refs [#10632](https://github.com/twisted/twisted/issues/10632)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_32"></a>|@therve removed owner|
|-|-|

It tried something with a decorator on the last commit. It does improve the situation a bit, but it's not that great either.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun set owner to @therve|
|-|-|

Some simple things:
  * FilePath.copyTo might be prettier than shutil.copy
  * Having the name ```testplugin``` as a symbolic constant (shared between ```setUp```, ```test_cache```, and ```test_deployedMode```) would make it more clear what the tests are trying to accomplish.
  * ```test_detectFilesRemoved``` docstring talks about a _removed plugin_ - really this case is testing the removal of a _dropin file_. (maybe the terminology is dumb, but that's what it is)
  * ```test_deployedMode``` docstring says _overriden_ but it probably means _overwritten_ or maybe even _updated_.
  * whitespace around the ```*``` on the modified ```os.utime``` line in ```DeveloperSetupTests.setUp}|}}

Overall, nice job.```{DeveloperSetupTests.setUp}|}}

Overall, nice job.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve removed owner|
|-|-|

Alright, that should be better!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun set owner to @therve|
|-|-|

Great, please merge

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_36"></a>|@therve set status to `closed`|
|-|-|

(In [21907]) Merge test-plugins-not-shared-942

Author: therve
Reviewers: exarkun, jml, glyph
Fixes [#10632](https://github.com/twisted/twisted/issues/10632)

Fix twisted.test.test_plugin so that it doesn't use twisted.plugins anymore,
but create temporary plugins beneath _trial_temp instead.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
